### PR TITLE
Update `sam --info` to display enriched info

### PR DESCRIPTION
**Pull request by hawflau**
#### Which issue(s) does this change fix?
N/A

#### Why is this change necessary?
To enrich `sam --info` to display system info. This makes it more straightforward for customers to provide diagnostic info when reporting bugs.

#### How does it address the issue?
In addition to `version`, `sam --info` now shows more info, e.g.:
```
{
  "version": "1.67.0",
  "system": {
    "python": "3.9.12",
    "os": "macOS-12.6.1-x86_64-i386-64bit"
  },
  "additional_dependencies": {
    "docker_engine": "20.10.12",
    "aws_cdk": "2.58.0 (build 4a577f3)",
    "terraform": "1.3.6"
  }
}
```

#### What side effects does this change have?
No

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [x] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/cli/main.py [hunk 1]
```diff
@@ -49,11 +50,115 @@ def print_info(ctx, param, value):
...
+    Gather system info
+
+    Returns
+    -------
+    dict[str, str]
+    """
+    from platform import platform, python_version
+
+    info = {
+        "python": python_version(),
```
Comment (moelasmar) [inline samcli/cli/main.py:74]: did you try to run `sam --info` for the different ways of installations, specially if there is no python installed on the machine?

**Review by moelasmar — DISMISSED** [review]
in general it looks good, but can we add some integration test case to validate that `sam --info` is working fine and does not break.

File: samcli/cli/main.py [hunk 2]
```diff
@@ -49,11 +50,115 @@ def print_info(ctx, param, value):
     if not value or ctx.resilient_parsing:
         return
 
-    click.echo(json.dumps({"version": __version__}, indent=2))
+    info = {
+        "version": __version__,
+        "system": _gather_system_info(),
+        "additional_dependencies": _gather_additional_dependencies_info(),
```
Comment (sriram-mv) [inline samcli/cli/main.py:56]: Might be better to move "atleast" additional deps to another file.

File: samcli/cli/main.py (hunk 1 above)
Comment (hawflau) [reply samcli/cli/main.py:74]: `sam` command is run on the Python interpreter installed along with SAM CLI. So it still works even if a user hasn't installed Python on their own.

Comment (hawflau) [reply samcli/cli/main.py:74]: Also confirmed we have existing integration test for `sam --info` already - https://github.com/aws/aws-sam-cli/blob/develop/tests/integration/root/test_root_command.py#L13

File: samcli/cli/main.py (hunk 2 above)
Comment (hawflau) [reply samcli/cli/main.py]: updated

File: samcli/lib/utils/system_info.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    Gather system info
+
+    Returns
+    -------
+    dict[str, str]
+    """
+    from platform import platform, python_version
+
+    info = {
+        "python": python_version(),
```
Comment (sriram-mv) [inline samcli/lib/utils/system_info.py:20]: in the case of MSI and pyinstaller, I'm assuming this gives back the bundled python version information?

**Review by sriram-mv — APPROVED** [review]

**Review by lucashuy — APPROVED** [review]

File: samcli/lib/utils/system_info.py (hunk 3 above)
Comment (hawflau) [reply samcli/lib/utils/system_info.py:20]: yes, that should be the python version of the python interpreter running the SAM CLI command.

**Pull request merged**